addons/wallet/models/sale_order.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, fields, models, _
+from odoo.tools import float_compare
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    # Relational fields
+    wallet_product_id = fields.Many2one('product.product', compute="_compute_wallet_product", store=False)
+    order_line_wallet_id = fields.Many2one('sale.order.line')
+    generated_gift_card_ids = fields.One2many('gift.card', "sale_order_id",
+                                              string="Bought Gift Card",
+                                              copy=False)
+
+    generated_gift_card_count = fields.Integer(compute="_compute_gift_card_purchased")
+
+    wallet_balance = fields.Monetary(compute='_compute_wallet_balance', string="Balance")
+    wallet_actual_balance = fields.Monetary(compute='_compute_wallet_actual_balance', string="Actual Balance")
+
+    # Compute Func
+    @api.depends("partner_id")
+    def _compute_wallet_balance(self):
+        for record in self:
+            wallet = record.get_wallet_by()
+            record.wallet_balance = wallet.currency_id._convert(
+                wallet.balance, record.currency_id,
+                self.env.company, fields.Date.today()
+            )
+
+    @api.depends("partner_id")
+    def _compute_wallet_actual_balance(self):
+        for record in self:
+            wallet = record.get_wallet_by()
+            record.wallet_actual_balance = wallet.currency_id._convert(
+                wallet.actual_balance, record.currency_id,
+                self.env.company, fields.Date.today()
+            )
+
+    def get_wallet_by(self):
+        return self.partner_id.get_wallet_by()
+
+    @api.constrains('state')
+    def _purchase_gift_card(self):
+        for record in self.filtered(lambda so: so.state == 'sale' and len(so.generated_gift_card_ids) == 0):
+            for gift_card_oder_line in record.order_line.filtered(lambda ol: ol.product_id.is_gift_card):
+                record._create_gift_cards(gift_card_oder_line)
+            record._send_gift_card_mail()
+
+    @api.depends("generated_gift_card_ids")
+    def _compute_gift_card_purchased(self):
+        for record in self:
+            record.generated_gift_card_count = len(record.generated_gift_card_ids)
+
+    @api.depends()
+    def _compute_wallet_product(self):
+        product = self.env['product.product'].search([('default_code', '=', 'PAY_WITH_WALLET')], limit=1)
```

Comment:
why not use a product define in data with an xml id ? 
and no reson to have wallet_product_id = fields in this case...

Reply:
If the user delete the product ? (it's what i did at first)

Reply:
check with @api.ondelete(at_uninstall=False)